plugins/cache-ehcache/src/main/java/org/opensearch/cache/keystore/RBMIntKeyLookupStore.java

```diff
@@ -0,0 +1,406 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.cache.keystore;
+
+import org.opensearch.common.cache.keystore.KeyLookupStore;
+import org.opensearch.common.metrics.CounterMetric;
+import org.opensearch.core.common.unit.ByteSizeValue;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+
+import org.roaringbitmap.RoaringBitmap;
+
+/**
+ * This class implements KeyLookupStore using a roaring bitmap with a modulo applied to values.
+ * The modulo increases the density of values, which makes RBMs more memory-efficient. The recommended modulo is ~2^28.
+ * It also maintains a hash set of values which have had collisions. Values which haven't had collisions can be
+ * safely removed from the store. The fraction of collided values should be low,
+ * about 0.5% for a store with 10^7 values and a modulo of 2^28.
```

Comment:
So collision will be dependent on the `hashCode` of key and the modulo chosen for store. Also I see that hashcode used here is essentially the hashcode of java key objects which I don't think has very low collision rate.

Reply:
Yes, the 0.5% number is just from the modulo. I didn't realize Java's hash codes had a high collision rate, but if that's a concern, for the new pluggable caches like EhcacheDiskCache we always use a wrapper object (ICacheKey) as the key. So we could change the hashcode function to whatever we like.